File: src/languages/madurese/madurese.ts
/**
 * Madurese / Bhâsa Madhurâ (mad) phonemizer — Austronesian, the 2008-revision Latin orthography, canonical IPA,
 * espeak-independent. NOT a direct grapheme map: Madurese has a signature VOWEL-REGISTER HARMONY (vowels raise
 * a→ɤ, ɛ→i, ə→ɨ, ɔ→u after a voiced OR aspirated stop; low elsewhere; ⟨l r w y⟩ transparent), plus glide/glottal
 * epenthesis between vowels, word-final devoicing, and geminate length. See
 * docs/investigations/mad_native_bringup_investigation.md.
 *
 * Pipeline: tokenize (greedy longest-match, tracking geminates) → epenthesis (insert ʔ between identical vowels,
 * w/j between different vowels unless height rises) → harmony (a register state machine sets each vowel's height)
 * → final devoicing → join.
 */
import type { Phonemizer } from "../../registry.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { numberToWords } from "./numbers.ts";
import { MANIFEST, CONS, CONS_KEYS, VOWEL_KEYS, type Reg } from "./manifest.ts";

const VSPEC = MANIFEST.vowelSpec;
const DEVOICE = MANIFEST.finalDevoice;
const CLAUSE_MARK = MANIFEST.clausePunctuation;

type Seg =
    | { kind: "C"; ipa: string; reg: Reg }
    | { kind: "V"; low: string; high: string };

const isConsLetter = (ch: string): boolean =>
    CONS.has(ch) || ch === "h"; // h forms digraphs (bh/dh/…) and is itself a consonant
const HEIGHT: Record<string, number> = { a: 0, ɛ: 1, ə: 1, ɔ: 1, e: 1, i: 2, u: 2 };
const isFront = (u: string): boolean => u === "i" || u === "ɛ" || u === "e";

/** Tokenize a word into consonant/vowel segments, expanding geminates to a length-marked consonant. */
function tokenize(w: string): Seg[] {
    const segs: Seg[] = [];
    let i = 0;
    let geminate = false;
    while (i < w.length) {
        // A doubled consonant LETTER marks a geminate: skip the first, length-mark the real consonant next round
        // (⟨tt⟩→tː, ⟨ss⟩→sː, ⟨bbh⟩→pʰː; liquids ⟨rr ll⟩ degeminate to a plain single).
        if (w[i] === w[i + 1] && isConsLetter(w[i]!)) {
            geminate = true;
            i += 1;
            continue;
        }
        let matched = false;
        for (const key of CONS_KEYS) {
            if (w.startsWith(key, i)) {
                const c = CONS.get(key)!;
                let ipa = c.ipa;
                if (geminate) {
                    if (c.ipa !== "l" && c.ipa !== "ɾ") ipa += "ː"; // liquids degeminate
                    geminate = false;
                }
                segs.push({ kind: "C", ipa, reg: c.reg });
                i += key.length;
                matched = true;
                break;
            }
        }
        if (matched) continue;
        for (const key of VOWEL_KEYS) {
            if (w.startsWith(key, i)) {
                const [low, high] = VSPEC[key]!;
                segs.push({ kind: "V", low, high });
                i += key.length;
                matched = true;
                break;
            }
        }
        if (!matched) {
            geminate = false;
            i += 1; // unknown char → skip
        }
    }
    return segs;
}

/** Insert epenthetic glides/glottal between adjacent vowels (Madurese vowel hiatus). */
function epenthesize(segs: Seg[]): Seg[] {
    const out: Seg[] = [];
    for (let k = 0; k < segs.length; k++) {
        out.push(segs[k]!);
        const a = segs[k]!;
        const b = segs[k + 1];
        // Decisions use the LOW (underlying) vowel quality — hiatus is resolved before the harmony sets height.
        if (a.kind === "V" && b && b.kind === "V") {
            if (a.low === b.low) out.push({ kind: "C", ipa: "ʔ", reg: "low" }); // identical → glottal
            else if ((HEIGHT[b.low] ?? 0) <= (HEIGHT[a.low] ?? 0))
                out.push({ kind: "C", ipa: isFront(a.low) ? "j" : "w", reg: "transp" }); // non-rising → glide
            // rising in height → true hiatus, no insertion
        }
    }
    return out;
}

/** Phonemize a single Madurese word to canonical IPA (register harmony + epenthesis + devoicing + gemination). */
export function phonemizeWord(word: string): string {
    const segs = epenthesize(tokenize(word.toLowerCase()));
    // A vowel's height is set by the register carried to it: a raise-class (voiced/aspirated) stop sets HIGH, a
    // low-class consonant sets LOW. A TRANSPARENT consonant (l r w y) is invisible to the harmony only inside an
    // ONSET CLUSTER — it passes a preceding stop's register through (bra→bɾɤ) — but a liquid AFTER a vowel is a
    // fresh onset and resets to low (bâla→bɤla). We approximate the cluster test as "the previous segment was a
    // consonant" (no syllabifier).
    let prevRaise = false; // word-initial = low
    let prevKind: "start" | "C" | "V" = "start";
    const ipa: string[] = [];
    let lastC = -1;
    for (const s of segs) {
        if (s.kind === "C") {
            if (s.reg === "raise") prevRaise = true;
            else if (s.reg === "low") prevRaise = false;
            else if (prevKind !== "C") prevRaise = false; // transparent liquid after a vowel/start → fresh low onset
            ipa.push(s.ipa);
            lastC = ipa.length - 1;
            prevKind = "C";
        } else {
            ipa.push(prevRaise ? s.high : s.low);
            lastC = -1; // a vowel is last, so no final-consonant devoicing
            prevKind = "V";
        }
    }
    // Word-final devoicing: a final voiced/aspirated stop neutralises to voiceless-unaspirated (length-insensitive,
    // so a final geminate ⟨…bb⟩→[pː] devoices too).
    if (lastC === ipa.length - 1 && lastC >= 0) {
        const seg = ipa[lastC]!;
        const long = seg.endsWith("ː");
        const d = DEVOICE[long ? seg.slice(0, -1) : seg];
        if (d) ipa[lastC] = long ? d + "ː" : d;
    }
    return ipa.join("");
}

// A word (Madurese letters incl. â è é ò ḍ ṭ and the ' glottal) / number / punctuation token.
const TOKEN = /([a-zâèéòḍṭ']+)|(\d+)|([.!?…,;:])/giu;

class MaduresePhonemizer implements Phonemizer {
    text(input: string): string {
        return assembleClauses(input, TOKEN, (m, sink) => {
            if (m[1]) sink.emit(phonemizeWord(m[1]));
            else if (m[2])
                for (const wd of numberToWords(Number(m[2])).split(" ")) sink.emit(phonemizeWord(wd));
            else if (m[3]) {
                const mk = CLAUSE_MARK[m[3]];
                if (mk) sink.pause(mk);
            }
        });
    }
}

/** Build the Madurese phonemizer (register-harmony rule g2p; tone/stress not salient). */
export function createMadurese(): Phonemizer {
    return new MaduresePhonemizer();
}

File: csharp/Vernacula.Phonemizer/Languages/Madurese/Madurese.cs
/**
 * Madurese / Bhâsa Madhurâ (mad) phonemizer — Austronesian, the 2008-revision Latin orthography, canonical
 * IPA. Not a grapheme map: vowel-register harmony, glide/glottal epenthesis, final devoicing, geminates.
 * Ported from src/languages/madurese/madurese.ts — see that file for the corpus evidence.
 */
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Madurese;

/** A tokenized segment: a consonant with its register class, or a vowel with its low/high realisations. */
public sealed class Seg
{
    public required bool IsVowel { get; init; }
    public string Ipa { get; set; } = "";
    public string RegClass { get; init; } = "";
    public string Low { get; init; } = "";
    public string High { get; init; } = "";
}

public static class MaduresePhonemizer
{
    private static IReadOnlyDictionary<string, string[]> VSPEC => Manifest.MANIFEST.VowelSpec;
    private static IReadOnlyDictionary<string, string> DEVOICE => Manifest.MANIFEST.FinalDevoice;
    private static IReadOnlyDictionary<string, string> CLAUSE_MARK => Manifest.MANIFEST.ClausePunctuation;

    /** TS `w.startsWith(key, i)`. */
    private static bool StartsWithAt(string w, string search, int i) =>
        i <= w.Length - search.Length && string.CompareOrdinal(w, i, search, 0, search.Length) == 0;

    private static bool IsConsLetter(string ch) => Manifest.CONS.ContainsKey(ch) || ch == "h";

    private static readonly IReadOnlyDictionary<string, double> HEIGHT = new Dictionary<string, double>(StringComparer.Ordinal)
    {
        ["a"] = 0, ["ɛ"] = 1, ["ə"] = 1, ["ɔ"] = 1, ["e"] = 1, ["i"] = 2, ["u"] = 2,
    };

    private static bool IsFront(string u) => u == "i" || u == "ɛ" || u == "e";

    private static double Height(string v) => HEIGHT.TryGetValue(v, out var h) ? h : 0;

    /** Tokenize a word into consonant/vowel segments, expanding geminates to a length-marked consonant. */
    private static List<Seg> Tokenize(string w)
    {
        var segs = new List<Seg>();
        var i = 0;
        var geminate = false;
        while (i < w.Length)
        {
            // TS indexes UTF-16 units (`w[i]`), and `w[i + 1]` past the end is `undefined` — never equal.
            var here = w[i].ToString();
            if (i + 1 < w.Length && w[i] == w[i + 1] && IsConsLetter(here))
            {
                geminate = true;
                i += 1;
                continue;
            }
            var matched = false;
            foreach (var key in Manifest.CONS_KEYS)
            {
                if (!StartsWithAt(w, key, i)) continue;
                var c = Manifest.CONS[key];
                var ipa = c.Ipa;
                if (geminate)
                {
                    if (c.Ipa != "l" && c.Ipa != "ɾ") ipa += "ː"; // liquids degeminate
                    geminate = false;
                }
                segs.Add(new Seg { IsVowel = false, Ipa = ipa, RegClass = c.RegClass });
                i += key.Length;
                matched = true;
                break;
            }
            if (matched) continue;
            foreach (var key in Manifest.VOWEL_KEYS)
            {
                if (!StartsWithAt(w, key, i)) continue;
                var spec = VSPEC[key];
                segs.Add(new Seg { IsVowel = true, Low = spec[0], High = spec[1] });
                i += key.Length;
                matched = true;
                break;
            }
            if (!matched)
            {
                geminate = false;
                var ph = LatinPhones.LatinPhone(here, new PhoneOpts { Initial = i == 0, IncludeH = true });
                if (ph is not null) segs.Add(new Seg { IsVowel = false, Ipa = ph, RegClass = Reg.Low });
                i += 1;
            }
        }
        return segs;
    }

    /** Insert epenthetic glides/glottal between adjacent vowels (Madurese vowel hiatus). */
    private static List<Seg> Epenthesize(List<Seg> segs)
    {
        var outp = new List<Seg>();
        for (var k = 0; k < segs.Count; k++)
        {
            outp.Add(segs[k]);
            var a = segs[k];
            var b = k + 1 < segs.Count ? segs[k + 1] : null;
            // Decisions use the LOW (underlying) vowel quality — hiatus resolves before harmony sets height.
            if (!a.IsVowel || b is null || !b.IsVowel) continue;
            if (a.Low == b.Low)
                outp.Add(new Seg { IsVowel = false, Ipa = "ʔ", RegClass = Reg.Low });
            else if (Height(b.Low) <= Height(a.Low))
                outp.Add(new Seg { IsVowel = false, Ipa = IsFront(a.Low) ? "j" : "w", RegClass = Reg.Transp });
            // rising in height → true hiatus, no insertion
        }
        return outp;
    }

    /** One Madurese word → canonical IPA (register harmony + epenthesis + devoicing + gemination). */
    public static string PhonemizeWord(string word)
    {
        var segs = Epenthesize(Tokenize(Js.ToLowerCase(word)));
        var prevRaise = false; // word-initial = low
        var prevKind = "start"; // "start" | "C" | "V"
        var ipa = new List<string>();
        var lastC = -1;
        foreach (var s in segs)
        {
            if (!s.IsVowel)
            {
                if (s.RegClass == Reg.Raise) prevRaise = true;
                else if (s.RegClass == Reg.Low) prevRaise = false;
                else if (prevKind != "C") prevRaise = false; // transparent liquid after a vowel → fresh onset
                ipa.Add(s.Ipa);
                lastC = ipa.Count - 1;
                prevKind = "C";
            }
            else
            {
                ipa.Add(prevRaise ? s.High : s.Low);
                lastC = -1; // a vowel is last, so no final-consonant devoicing
                prevKind = "V";
            }
        }
        if (lastC == ipa.Count - 1 && lastC >= 0)
        {
            var seg = ipa[lastC];
            var lng = seg.EndsWith("ː", StringComparison.Ordinal);
            // TS `if (d)`: an absent key AND an empty-string mapping both leave the segment alone.
            if (DEVOICE.TryGetValue(lng ? seg[..^1] : seg, out var d) && d != "")
                ipa[lastC] = lng ? d + "ː" : d;
        }
        return string.Concat(ipa);
    }

    // A word (Madurese letters incl. â è é ò ḍ ṭ and the ' glottal) / number / punctuation token.
    private static readonly JsRe TOKEN =
        JsRegex.Compile($"({HostWord.HostWordRun(new[] { "Latin" }, "'")})|(\\d+)|([.!?…,;:])", "giu");

    /** This language's OWN inventory — a token this class rejects carries a letter Madurese does not use. */
    private const string NATIVE_CLASS = "[a-zâèéòḍṭ']";
    private static readonly Func<string, string> Nat = HostWord.MakeNativiser(NATIVE_CLASS, "iu");

    private sealed class Engine : ILanguage
    {
        public string Text(string input)
        {
            // ⚠ NORMALIZATION RUNS FIRST: the tokenizer splits on a bare `\d+` and treats `.`/`,` as clause
            // punctuation, so a grouping separator or a decimal comma would become a PHRASE BREAK.
            return Clauses.AssembleClauses(Normalize.NormalizeMadurese(input), TOKEN, (m, sink) =>
            {
                if (m.Groups[1].Success && m.Groups[1].Value.Length > 0)
                    sink.Emit(PhonemizeWord(Nat(m.Groups[1].Value)));
                else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0)
                {
                    var digits = m.Groups[2].Value;
                    foreach (var wd in Numbers.NumberToWords(Js.Number(digits), digits).Split(' '))
                        sink.Emit(PhonemizeWord(wd));
                }
                else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0)
                {
                    var mk = CLAUSE_MARK.GetValueOrDefault(m.Groups[3].Value);
                    if (mk is not null) sink.Pause(mk);
                }
            });
        }
    }

    /** Build the Madurese phonemizer (register-harmony rule g2p; tone/stress not salient). */
    public static ILanguage CreateMadurese() => new Engine();

    internal static void RegisterSelf() => Registry.Register("madurese", CreateMadurese);
}
